src/classlibnative/float/floatdouble.cpp

```diff
@@ -0,0 +1,300 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+// 
+// File: FloatDouble.cpp
+// 
+
+#include <common.h>
+
+#include "floatdouble.h"
+#include "floatclass_double.h"
+
+#define AS_BITFIELD(x)         (*((UINT64*)(&x)))
+
+#define IS_INFINITY(x)         ((AS_BITFIELD(x) & UI64(0x7FFFFFFFFFFFFFFF)) == UI64(0x7FF0000000000000))
+#define IS_ONE(x)              (AS_BITFIELD(x) == UI64(0x3FF0000000000000))
+#define IS_NEGATIVEONE(x)      (AS_BITFIELD(x) == UI64(0xBFF0000000000000))
+
+#ifdef _MSC_VER
+#pragma float_control(precise, off)
+#endif
+
+// Trigonometric Functions
+FCIMPL1_V(double, COMDouble::Acos, double x)
+    FCALL_CONTRACT;
+
+    return acos(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Asin, double x)
+    FCALL_CONTRACT;
+
+    return asin(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Atan, double x)
+    FCALL_CONTRACT;
+
+    return atan(x);
+FCIMPLEND
+
+FCIMPL2_VV(double, COMDouble::Atan2, double y, double x)
+    FCALL_CONTRACT;
+
+    // The cmath::atan2 method does not produce nan for atan2(+-inf,+-inf)
+    return (IS_INFINITY(y) && IS_INFINITY(x)) ? (y / x) : atan2(y, x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Cos, double x)
+    FCALL_CONTRACT;
+
+    return cos(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Sin, double x)
+    FCALL_CONTRACT;
+
+    return sin(x);
+FCIMPLEND
+
+FCIMPL2_VI(double, COMDouble::SinCos, double x, double* y)
+    FCALL_CONTRACT;
+
+    *y = COMDouble::Cos(x);
+    return COMDouble::Sin(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Tan, double x)
+    FCALL_CONTRACT;
+
+    return tan(x);
+FCIMPLEND
+
+// Hyperbolic Functions
+FCIMPL1_V(double, COMDouble::Acosh, double x)
+    FCALL_CONTRACT;
+
+    return acosh(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Asinh, double x)
+    FCALL_CONTRACT;
+
+    return asinh(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Atanh, double x)
+    FCALL_CONTRACT;
+
+    return atanh(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Cosh, double x)
+    FCALL_CONTRACT;
+
+    return cosh(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Sinh, double x)
+    FCALL_CONTRACT;
+
+    return sinh(x);
+FCIMPLEND
+
+FCIMPL1_V(double, COMDouble::Tanh, double x)
+    FCALL_CONTRACT;
+
+    return tanh(x);
+FCIMPLEND
+
+// Exponential Functions
+FCIMPL1_V(double, COMDouble::Exp, double x)
+    FCALL_CONTRACT;
+
+    // the cmath::exp method does not handle +-inf properly
+    return IS_INFINITY(x) ? ((x < 0) ? +0.0 : x) : exp(x);
```

Comment:
This is no longer the case in MSVC 15.

`cmath::exp` now returns `+0.0` for negative infinity and positive infinity for positive infinity.